### DISCUSSION TITLE
How can I refresh the valuesets manually

### DISCUSSION THREAD

**Discussion by HausBJB** [DevTrevi/DgcReader#32] [Q&A] [answered]
Hello, I have a question about refresh ValueSets. I can manually refresh the Trust List and also the Rules but not the ValueSets ?? the valuesets are loaded at the first validation ?? correct? therefore the first query is slower. there is a possibility to update it manually

Thanks for the great work, the library is beautiful

**Comment by DevTrevi** [comment]
Hi @HausBJB, thank you :)

There are currently no methods at the moment to manually refresh valuesets for the GermanRulesProvider.
I think that I could call the refresh methods for all the valuesets in the `RefreshRules` method, or maybe I could also add some dedicated methods to update every single valueset.
I will add these methods in the next release.

Please note that the GermanRulesValidator is not an official implementation, and may not be fully compliant with current regulations

**Comment by HausBJB** [comment]
Thank you for your prompt reply. And that would be great if you can refresh everything manually. The rules for Germany are loaded from the same server as with the covpass app. The library works well and I see no reason not to use it. or what do you mean by "and may not fully comply with applicable regulations"

**Comment by DevTrevi** [comment]
Hi @HausBJB 

I confirm that rules are downloaded by the same server as the official [Digitaler-Impfnachweis covpass-android app](https://github.com/Digitaler-Impfnachweis/covpass-android/), but my implementation of the engine for these rules has not been validated by any official authority, so I can not guarantee that is bug free nor that will fully cover all the use cases of the official sdk.
I made the `DgcReader.RuleValidators.Germany` implementation primarly as a proof of concept, in order to demonstrate the extensibility of the library.

For instance, the official sdk has some logic to warn users that needs a booster dose, but I have not implemented that functionality in the library. Also the 2G+ restrictions are not managed, as it is not clear to me how they apply to each country supported by the official Digitaler-Impfnachweis sdk (it seems that even the official sdk does not cover this use case at the moment).

What I'm saying is that if you use this library, it's your own responsibility to verify if regulations allows you to implement an application that uses an sdk different from the official one. 
As stated in the readme, I will not take any responsibility for any consequence about the usage of this library, so if your country requires you to use the official sdk only, you will not be compliant when using this library.

I just wanted to make this clear, to avoid misunderstandings :)

**Comment by HausBJB** [comment]
Hi it is understandable that they are not responsible for it. I want to use that for 3G at work. I don't know yet whether that's legally ok. Since the library can be expanded, the data from the certificates are compared with the personal data. I think the booster rules are only intended for notification and not to validate it. How is it that in Italy is it allowed to use what is your own or you are forced to use the official one. Thanks again for the great work.

**Reply by DevTrevi** [reply]
Italian authorities opened to 3rd parties SDKs by allowing the submission of libraries to be certified (see [the official sdk onboarding repo](https://github.com/ministero-salute/it-dgc-verificac19-sdk-onboarding))
The Italian implementations in this repository has been validated (`DgcReader`, `DgcReader.TrustListProviders.Italy`, `DgcReader.RuleValidators.Italy`, `DgcReader.BlacklistProviders.Italy`) by Italian authorities, so you can legally use them for validating certificates using "IT" as acceptance country

**Comment by DevTrevi** [comment, marked as answer]
Hi @HausBJB , I just released the v2.1.0 that now exposes two methods for manually refreshing rules and valuesets.
If you call the standard RefreshRules method without specifying the country code, all rules and valueset will be updated
See this commit for details #39

**Comment by HausBJB** [comment]
Thank you.

that was fast
